### ISSUE TITLE
Path to build artifacts changes depending on the build-type

### ISSUE THREAD

**Issue by 23Skidoo** [haskell/cabal#4771]
For example, if I have a package `foo` with an executable `exe`, with `build-type: Simple` the executable will be under

```
dist-newstyle/build/x86_64-linux/ghc-8.2.1/foo-1.0/c/exe/build/exe/exe
```

But if I change the build type to `Custom`, it will be under

```
dist-newstyle/build/x86_64-linux/ghc-8.2.1/foo-1.0/build/exe/exe
```

This is with Cabal 2.0.

/cc @ezyang

**Comment by ezyang** [comment]
Unfortunately, this is expected behavior. In particular, it has less to do with Custom, and more to do with whether or not we support per-component builds; per-component builds have a different directory layout.

Is there a broader context where this is a problem?

**Comment by 23Skidoo** [comment]
It's just a bit confusing and annoying. For example, if a user switches a `build-type: Simple` project to `build-type: Custom`, it may take them a while to realise that the directory layout is now different. Plus if those paths are hardcoded somewhere (e.g. in CI conf, scripts or an external build system) they now have to be changed.

**Comment by ezyang** [comment]
So, there are some workarounds available to us. For example, after running an old-style setup script, we could copy / otherwise move binaries so that they show up in the locations people "expect". This might be a bit too magical, but it would "solve" the problem, to a certain extent.

**Comment by ivan-m** [comment]
Is there support to programmatically discover where these paths are? If that's promoted, then scripts can refer to those locations rather than trying to discover them themselves.

**Comment by gbaz** [comment]
Agreed -- a flag to builds that output paths to executable artifacts as a result would be welcome in general.

**Comment by hvr** [comment]
@ivan-m for new-build, you can use http://hackage.haskell.org/package/cabal-plan to programmatically query the paths for build targets.
